cpp/include/graph_generators.hpp

```diff
@@ -0,0 +1,314 @@
+ * @param size_distribution Distribution of the graph sizes, impacts the scale parameter of the
+ * R-MAT generator
+ * @param edge_distribution Edges distribution for each graph, impacts how R-MAT parameters a,b,c,d,
+ * are set.
+ * @param seed Seed value for the random number generator.
+ * @param clip_and_flip Flag controlling whether to generate edges only in the lower triangular part
+ * (including the diagonal) of the graph adjacency matrix (if set to `true`) or not (if set to
+ * `false`).
+ * @param scramble_vertex_ids Flag controlling whether to scramble vertex ID bits (if set to `true`)
+ * or not (if set to `false`); scrambling vertx ID bits breaks correlation between vertex ID values
+ * and vertex degrees. The scramble code here follows the algorithm in the Graph 500 reference
+ * implementation version 3.0.0.
+ * @return A vector of std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>> of
+ *size @p n_edgelists, each vector element being a tuple of rmm::device_uvector objects for edge
+ *source vertex IDs and edge destination vertex IDs.
+ */
+template <typename vertex_t>
+std::vector<std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>>>
+generate_rmat_edgelists(
+  raft::handle_t const& handle,
+  size_t n_edgelists,
+  size_t min_scale,
+  size_t max_scale,
+  size_t edge_factor                         = 16,
+  generator_distribution_t size_distribution = generator_distribution_t::POWER_LAW,
+  generator_distribution_t edge_distribution = generator_distribution_t::POWER_LAW,
+  uint64_t seed                              = 0,
+  bool clip_and_flip                         = false,
+  bool scramble_vertex_ids                   = false);
+
+/**
+ * @brief generate an edge list for randomly ordered path graph
+ *
+ * Given @p num_vertices, the vertices are randomly ordered and a set of
+ * @p num_vertices - 1 edges is constructed visiting each vertex in this
+ * random order exactly once.
+ *
+ * If executed in a multi-gpu context (handle comms has been initialized)
+ * the path will span all GPUs
+ *
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param num_vertices Number of vertices to use in generation
+ * @param symmetrize If true, symmetrize the edges
+ * @param seed Seed value for the random number generator.
+ * @return std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>> A tuple of
+ * rmm::device_uvector objects for edge source vertex IDs and edge destination vertex IDs.
+ */
+template <typename vertex_t>
+std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>>
+generate_random_path_graph_edgelist(raft::handle_t const& handle,
+                                    size_t num_vertices,
+                                    bool symmetrize = false,
```

Comment:
We should decide how to handle symmetrize. Currently, R-mat generator does not symmetrize but if clip_and_flip is set to true, generates edges only in the lower triangular (including the diagonal) area of the graph adjacency matrix.

If we are willing to symmetrize in edge list, we may add symmetrize to R-mat generator public API (and remove clip_and_flip) and R-mat generator will call internal API with clip_and_flip and for all non-diagonal edges, create a new edge flipping the source and destination.

Another option is to symmetrize after graph creation. This will be more complex to implement, but we can limit the maximum memory usage with this approach.

Reply:
Actually, we should not symmetrize R-mat output here. It may not be acceptable in Graph 500 benchmarking.

I think symmetrizing in graph creation or after graph creation will lead to better design in long run.

Reply:
I put symmetrize here because it was easy to implement at this point without worrying about having to dedupe edges.

However, after implementing some of these routines, it turns out for anything sophisticated we're going to have to dedupe edges anyway.  It feels like a separate symmetrize function would be more flexible.